src/java.base/share/classes/java/math/BigInteger.java

```diff
@@ -2701,6 +2693,85 @@ public BigInteger pow(int exponent) {
         }
     }
 
+    /**
+     * Computes {@code x^n} using repeated squaring trick.
+     * Assumes {@code x != 0 && x^n < 2^Long.SIZE}.
+     */
+    static long unsignedLongPow(long x, int n) {
+        if (x == 1L)
+            return 1L;
+
+        if (x == 2L)
+            return 1L << n;
+        // Double.PRECISION / bitLength(x) is the largest integer e
+        // such that x^e fits into a double. If e <= 3, we won't use fp arithmetic.
+        // This allows to use fp arithmetic in computePower(), where possible.
+        final int maxExp = Math.max(3, Double.PRECISION / bitLengthForLong(x));
+        final int maxExpLen = bitLengthForInt(maxExp);
+        final long[] powerCache = new long[1 << maxExpLen];
+        powerCache[1] = x;
+        powerCache[2] = x * x;
+        powerCache[3] = x * x * x;
+
+        final int nZeros = Integer.numberOfLeadingZeros(n);
+        n <<= nZeros;
+
+        long pow = 1L;
+        int blockLen;
+        for (int nLen = Integer.SIZE - nZeros; nLen > 0; nLen -= blockLen) {
+            blockLen = Math.min(maxExpLen, nLen);
+            // compute pow^(2^blockLen)
+            if (pow != 1L) {
+                for (int i = 0; i < blockLen; i++)
+                    pow *= pow;
+            }
+
+            // add exp to power's exponent
+            int exp = n >>> -blockLen;
+            if (exp > 0)
+                pow *= computePower(powerCache, x, exp, maxExp);
+
+            n <<= blockLen; // shift to next block of bits
+        }
+
+        return pow;
```

Comment:
Is there a reason this cannot simply be the traditional "repeated square" method?
Why this complexity?

```suggestion
        if (x == 1L)
            return 1L;

        if (x == 2L)
            return 1L << n;

        /*
         * The method assumption means that n <= 40 here.
         * Thus, the loop body executes at most 5 times.
         */ 
        long pow = 1;
        for (; n > 1; x *= x, n >>>= 1) {
            if ((n & 0b1) != 0) {
                pow *= x;
            }
        }
        return pow * x;
```

Reply:
@rgiulietti Probably the `powerCache` array is needless because the number of iterations is very low. Perhaps using `Math.pow(int, int)` could be a faster way to get the powers, since reduces the number of iterations further and it has constant time, although even the traditional "repeated square" method has constant time...

Reply:
@rgiulietti More than else, since `Math.pow(double, double)` has a lot of optimizations, I expect it to be faster than the simple "repeated square" method...

Reply:
@fabioromano1 pow(double,double) is quite complex, regardless of whether it is implemented in software or hardware. In fact, it's one of the most complex math functions, one of the reasons being that it takes two arguments.
The implementation invoked in `StrictMath` is more than 200 lines long (including comments).

The "repeated square" slow-path executes ≤ 11 `long` multiplications.
Plus, we need to balance performance and readability.

Reply:
> Is there a reason this cannot simply be the traditional "repeated square" method? Why this complexity?

@rgiulietti By the way, could make sense recycling this algorithm scheme to speed up the computing of powers of `int`s that don't fit into a `long`, using `unsignedLongPow(long, int)` instead of `Math.pow()`?